elki/src/main/java/de/lmu/ifi/dbs/elki/math/statistics/dependence/MCDEDependenceMeasure.java

```diff
@@ -0,0 +1,173 @@
+package de.lmu.ifi.dbs.elki.math.statistics.dependence;
+
+
+import de.lmu.ifi.dbs.elki.utilities.datastructures.arraylike.NumberArrayAdapter;
+import de.lmu.ifi.dbs.elki.utilities.documentation.Reference;
+import de.lmu.ifi.dbs.elki.utilities.exceptions.AbortException;
+import de.lmu.ifi.dbs.elki.utilities.random.RandomFactory;
+import de.lmu.ifi.dbs.elki.utils.containers.RankStruct;
+import java.util.Arrays;
+import java.util.Random;
+
+
+/**
+ * Implementation of bivariate Monte Carlo Density Estimation as described in
+ * Edouard Fouché & Klemens Böhm<br>
+ * Monte Carlo Density Estimation<br>
+ * Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)
+ *
+ * This is an abstract class. In order to use MCDE extend it and implement an appropriate statistical test that
+ * returns a p-value and index structure for efficient computation of the statistical test.
+ *
+ * The instantiation of MCDE based on the Mann-Whitney U test is called MWP (as described in the paper).
+ * See McdeMwpDependenceMeasure.
+ *
+ * @author Alan Mazankiewicz
+ * @author Edouard Fouché
+ */
+
+@Reference(authors = "Edouard Fouché, Klemens Böhm", //
+        title = "Monte Carlo Density Estimation", //
+        booktitle = "Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)",
+        url = "https://doi.org/10.1145/3335783.3335795", //
+        bibkey = "DBLP:conf/ssdbm/FoucheB19")
+
+
+public abstract class MCDEDependenceMeasure<R extends RankStruct> extends AbstractDependenceMeasure {
+    /**
+     * Monte-Carlo iterations.
+     */
+    protected int m = 50;
+
+    /**
+     * Expected share of instances in slice (independent dimensions).
+     */
+    protected double alpha = 0.5;
+
+    /**
+     * Share of instances in marginal restriction (reference dimension).
+     * Note that in the original paper alpha = beta and as such there is no explicit distinction between the parameters.
+     */
+    protected double beta = 0.5;
+
+    /**
+     * Random generator.
+     */
+    protected RandomFactory rnd;
+
+    /**
+     * Constructor.
+     *
+     * @param m Monte-Carlo iterations
+     * @param alpha Expected share of instances in slice (under independence)
+     * @param beta Share of instances in marginal restriction (reference dimension)
+     * @param rnd Random source
+     */
+    public MCDEDependenceMeasure(int m, double alpha, double beta, RandomFactory rnd){
+        if((beta > 1.0) || beta <= 0.0) throw new AbortException("beta must be in ]0;1]");
+        if((alpha >= 1.0) || alpha <= 0.0) throw new AbortException("alpha must be in ]0;1[ ");
```

Comment:
The if condition is better phrased as `!(alpha > 0. && alpha < 1.0)` because of NaN values.